### PR TITLE
Add SDL3 to CI and upload build artifacts

### PR DESCRIPTION
**Pull request by thatcosmonaut**
It would be nice to figure out how to spit out a universal dylib for MacOS, but I'm not familiar enough with how that works. It would also be nice to figure out how to upload the release artifacts to the release page automatically when a release is tagged, but I'm not sure how to test that.

### CODE REVIEW

**Comment by flibitijibibo** [comment]
In addition to macOS we also need a Rocky 8 VM for Linux builds - we should ask Misa about this.

**Comment by thatcosmonaut** [comment]
Something like this? https://www.claudiokuenzler.com/blog/1381/how-to-run-github-ci-cd-workflows-actions-rocky-linux

**Comment by flibitijibibo** [comment]
Oh, that looks a lot easier than what I remember... yeah, if that works with 8 then we can integrate that and the output binaries should be good to go.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -166,3 +180,168 @@ jobs:
           sudo pkg install -y cmake ninja sdl2
           cmake -B build -G Ninja .
           cmake --build build --verbose -- -j`sysctl -n hw.ncpu`
+
+  linux-mingw:
```
Comment (flibitijibibo) [inline .github/workflows/ci.yml:184]: Do we want to replace the Windows MinGW build with this? The only difference should be the host OS, so let's use whichever one is faster and drop the other.

**Review by flibitijibibo — CHANGES_REQUESTED** [review]
Output looks okay, so the final checklist:

- [x] MinGW (see comment)
- [x] Rocky Linux 8 x86_64
- [ ] Rocky Linux 8 aarch64

**Comment by thatcosmonaut** [comment]
https://github.blog/changelog/2024-09-03-github-actions-arm64-linux-and-windows-runners-are-now-generally-available/

>  We expect to begin offering Arm runners for open source projects by the end of the year.

Once this is enabled we can do the aarch builds. Working on the other stuff now.

**Review by flibitijibibo — COMMENTED** [review]
For MinGW, do the Rocky tools include `mingw32-cmake` and `mingw64-cmake`? That'll help eliminate the toolchain files if so.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -2,41 +2,118 @@ name: CI
...
+    - name: Clone SDL2
+      run: git clone --depth 1 --branch release-$SDL2_VERSION https://github.com/libsdl-org/SDL.git
+
+    - name: Build SDL2 (Release)
       run: |
-        sudo apt-add-repository ppa:hrzhu/sdl2-backport
-        sudo apt-get update
-        sudo apt-get install ninja-build libsdl2-dev
+        cd SDL
+        cmake -B release -G Ninja . -DCMAKE_BUILD_TYPE=Release
```
Comment (flibitijibibo) [inline .github/workflows/ci.yml:39]: We can trim down the build time with debug mode I bet

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -2,41 +2,118 @@ name: CI
...
+        with:
+          submodules: true
+
+      - name: Download SDL
+        run: |
+          curl -LO https://github.com/libsdl-org/SDL/releases/download/release-$SDL2_VERSION/SDL2-devel-$SDL2_VERSION-mingw.tar.gz
+          tar -xvf SDL2-devel-$SDL2_VERSION-mingw.tar.gz -C ${GITHUB_WORKSPACE}
+
+      - name: CMake configure 32-bit (Release)
+        run: cmake -B release-x86 -G Ninja . -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=.github/mingw-w64-i686.cmake -DSDL2_LIBRARIES=${GITHUB_WORKSPACE}/SDL2-$SDL2_VERSION/i686-w64-mingw32/bin/SDL2.dll -DSDL2_INCLUDE_DIRS=${GITHUB_WORKSPACE}/SDL2-$SDL2_VERSION/i686-w64-mingw32/include/SDL2
```
Comment (flibitijibibo) [inline .github/workflows/ci.yml:86]: I _think_ this still works...

```suggestion
        run: mingw32-cmake -B release-x86 -G Ninja . -DCMAKE_BUILD_TYPE=Release  -DSDL2_LIBRARIES=${GITHUB_WORKSPACE}/SDL2-$SDL2_VERSION/i686-w64-mingw32/bin/SDL2.dll -DSDL2_INCLUDE_DIRS=${GITHUB_WORKSPACE}/SDL2-$SDL2_VERSION/i686-w64-mingw32/include/SDL2
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -2,41 +2,118 @@ name: CI
...
+
+      - name: Build 32-bit (Release)
+        run: ninja -C release-x86
+
+      - name: CMake configure 64-bit (Release)
+        run: cmake -B release-x64 -G Ninja . -DCMAKE_BUILD_TYPE=Release -DCMAKE_TOOLCHAIN_FILE=.github/mingw-w64-x86_64.cmake -DSDL2_LIBRARIES=${GITHUB_WORKSPACE}/SDL2-$SDL2_VERSION/x86_64-w64-mingw32/bin/SDL2.dll -DSDL2_INCLUDE_DIRS=${GITHUB_WORKSPACE}/SDL2-$SDL2_VERSION/x86_64-w64-mingw32/include/SDL2
```
Comment (flibitijibibo) [inline .github/workflows/ci.yml:92]: Ditto:

```suggestion
        run: mingw64-cmake -B release-x64 -G Ninja . -DCMAKE_BUILD_TYPE=Release -DSDL2_LIBRARIES=${GITHUB_WORKSPACE}/SDL2-$SDL2_VERSION/x86_64-w64-mingw32/bin/SDL2.dll -DSDL2_INCLUDE_DIRS=${GITHUB_WORKSPACE}/SDL2-$SDL2_VERSION/x86_64-w64-mingw32/include/SDL2
```

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -77,92 +158,185 @@ jobs:
...
+
+    - name: Install dependencies
+      run: |
+        dnf update -y
+        dnf install -y epel-release
+        dnf config-manager --set-enabled powertools
+        dnf install -y bzip2 bzip2-devel chrpath clang cmake freetype-devel gcc-c++ \
+            git gtk3-devel libstdc++-static libuuid-devel libvorbis-devel \
+            libxml2-devel meson ninja-build ocl-icd-devel openal-soft-devel \
+            opencl-headers openssl-devel patch perl-IPC-Cmd svn unix2dos xz zlib-static
```
Comment (flibitijibibo) [inline .github/workflows/ci.yml:222]: Cleaning up my own mess:

```suggestion
        dnf install -y chrpath cmake gcc-c++ git ninja-build
```

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -2,41 +2,118 @@ name: CI
...
```
Comment (flibitijibibo) [inline .github/workflows/ci.yml:27]: Ditto:

```suggestion
        dnf install -y chrpath cmake gcc-c++ git ninja-build
```

**Review by flibitijibibo — CHANGES_REQUESTED** [review]
A few fixes here and there for the Rocky config - if this all works then the toolchain files can be deleted. We should also `chrpath -d` the Linux binaries, and possibly `strip -S` as well (not sure why this changes the Release binaries, but it does...)

**Comment by thatcosmonaut** [comment]
Rocky builds should be good now, it would be nice to spit out a universal MacOS dylib but I'm not gonna have time to work on that until next week at least.

**Review by flibitijibibo — APPROVED** [review]
Latest should be good as long as macOS cooperates...

**Pull request merged**